tests/functional/test_install.py

```diff
@@ -2371,14 +2372,68 @@ def test_install_logs_pip_version_in_debug(
     assert_re_match(pattern, result.stdout)
 
 
-def test_install_dry_run(script: PipTestEnvironment, data: TestData) -> None:
-    """Test that pip install --dry-run logs what it would install."""
-    result = script.pip(
-        "install", "--dry-run", "--find-links", data.find_links, "simple"
-    )
+@pytest.fixture
+def install_find_links(
+    script: PipTestEnvironment,
+    data: TestData,
+) -> Callable[[Iterable[str], bool, Optional[Path]], TestPipResult]:
+    def install(
+        args: Iterable[str], dry_run: bool, target_dir: Optional[Path]
+    ) -> TestPipResult:
+        return script.pip(
+            "install",
+            *(
+                (
+                    "--target",
+                    str(target_dir),
+                )
+                if target_dir is not None
+                else ()
+            ),
+            *(("--dry-run",) if dry_run else ()),
+            "--no-index",
+            "--find-links",
+            data.find_links,
+            *args,
+        )
+
+    return install
+
+
+@pytest.mark.parametrize(
+    "with_target_dir",
+    (True, False),
+)
+def test_install_dry_run_nothing_installed(
+    script: PipTestEnvironment,
+    tmpdir: Path,
+    install_find_links: Callable[[Iterable[str], bool, Optional[Path]], TestPipResult],
+    with_target_dir: bool,
+) -> None:
+    """Test that pip install --dry-run logs what it would install, but doesn't actually
+    install anything."""
+    if with_target_dir:
+        install_dir = tmpdir / "fake-install"
+        install_dir.mkdir()
+    else:
+        install_dir = None
+
+    result = install_find_links(["simple"], True, install_dir)
```

Comment:
Not particularly relevant since this is a test but it’d probably more readable to do something like `target=install_dir` to make it clearer it’s doing a `--target` install.

Reply:
I made `dry_run` and `target_dir` into keyword-only arguments in 674a2b7 which I agree makes it a little clearer. It would be preferable to explicitly specify `target_dir=None` or something but I thought that was a little more difficult to read than just passing `target_dir=install_dir` and propagating the `None`.

Reply:
Ugh, looks like `mypy_extensions` isn't available for any version less than 3.12, which means we can't call factory functions from fixtures like this with keyword arguments until then. We could use a `Protocol` with a `__call__` method, but that seems like overkill as you said for a test method. Reverting again, sadly.